.docs/architecture/adr-001-general-parser-architecture.md

```diff
@@ -0,0 +1,234 @@
+
+```go
+type Context interface {
+    // Context represents the manager for the concurrent jobs
+    context.Context
+
+    // Database returns the database used to operate blocks by workers
+    Database() Database
+
+    // Node returns the client used to listen to blocks by enqueuers
+    Node() Node
+    
+    // Modules returns a set of the extensional operations that should be executed inside a worker
+    Modules() []Module
+}
+```
+
+### Block
+
+The Block interface serves as a representation of the aggregation of operations on the blockchain and is the fundamental element transferred within the queue for parsing by workers within Juno.
+
+```go
+type Block interface {
+    // Height returns the height of the block.
+    Height() uint64
+
+    // Hash returns the hash of the block.
+    Hash() string
+
+    // Timestamp returns the timestamp of the block.
+    Timestamp() time.Time
+
+    // Proposer returns the address of the block proposer.
+    Proposer() string
+}
+```
+
+### Enqueuer
+
+The Enqueuer interface orchestrates the process of listening for available blocks and enqueueing them into the queue for parsing by workers.
+
+In the current version, two types of enqueuers are implemented:
+
+* Missing Blocks Enqueuer: This enqueuer queries missing blocks from the given start block to the current latest block and enqueues them into the queue.
+* New Blocks Enqueuer: This enqueuer listens for new blocks from the node and enqueues them into the queue.
+
+```go
+type Enqueuer interface {
+    // ListenAndEnqueueBlocks listens for available blocks and enqueues them into the queue.
+    ListenAndEnqueueBlocks(ctx Context, queue BlockQueue) (err error)
+}
+```
+
+### Block Queue
+
+The BlockQueue interface serves as a communication channel between enqueuers and workers within the Juno architecture.
+
+```go
+type BlockQueue interface {
+    // Enqueue enqueues a block into the queue.
+    Enqueue(block Block)
+    
+    // Listen returns a channel to listen for blocks.
+    Listen() <-chan Block
+}
+```
+
+### Node
+
+The Node interface serves as an abstraction layer for interacting with RPC servers or other clients to obtain block data efficiently.
+
+```go
+type Node interface {
+    // LatestBlock retrieves the latest block from the connected node.
+    LatestBlock() (Block, error)
+
+    // Block retrieves the block by the given height
+    Block(height uint64) (Block, error)
+
+    // SubscribeBlocks returns a channel to subscribe to new blocks from the connected node.
+    SubscribeBlocks() (<-chan Block, error)
+}
+```
+
+### Database
+
+The Database interface encapsulates the necessary operations for interacting with the underlying storage system.
+
+```go
+type Database interface {
+    // SaveBlock saves a block into the database.
+    SaveBlock(block Block) error
+
+    // LatestBlock returns the latest block from the database.
+    LatestBlock() (Block, error)
+}
```

Comment:
Are we planning on supporting multiple databases? 

Going Hexagonal/Clean architecture is sometimes detrimental in terms of performance and code maintainability. In my past experiences, I have seen multiple instances where generalization/composition makes this 10 times harder to reason and maintain.

I am not saying that what you propose is not good. I'm just trying to understand the benefit of an interface rather than the concrete instance.

Reply:
> Are we planning on supporting multiple databases?

This interface can let allow us try different database from my side. As we highly may not use PostgresSQL, and possible try with ScyllaDB or other NoSQL database.

> In my past experiences, I have seen multiple instances where generalization/composition makes this 10 times harder to reason and maintain.

I agree with you, however, we need a abstract layer as we gonna support different chains, so to trade off it, I think we should make interfaces as simple as possible.

Besides, I would introduce another benefit. Currently, we cannot test worker, enqueuer and many factors since Juno does not apply dependency injection.  Having these interfaces would help Juno to be more testable as well.

Reply:
> This interface can let allow us try different database from my side. As we highly may not use PostgresSQL, and possible try with ScyllaDB or other NoSQL database.

That's my concern. If we plan on supporting multiple databases we have to acknowledge that they might be fundamentally different. Apart from writing a good amount of data mapping code and the performance overhead, we will also prevent ourselves from using some features of each specific database. I believe that's why there's no popular ORM that can do SQL and NoSQL.

> I agree with you, however, we need a abstract layer as we gonna support different chains, so to trade off it, I think we should make interfaces as simple as possible.

I think we want to support different inputs and conform them to the same output. I think your approach works for Blocks but not for chain-specific elements. Let's say chains A and B have different proposal data structures, how can we accommodate the `Database` interface to persist the different shapes of the proposal? 

> Besides, I would introduce another benefit. Currently, we cannot test worker, enqueuer and many factors since Juno does not apply dependency injection. Having these interfaces would help Juno to be more testable as well.

That's actually a great point, are we not able to mock/spy/stub the Database client?

Reply:
> That's my concern. If we plan on supporting multiple databases we have to acknowledge that they might be fundamentally different. Apart from writing a good amount of data mapping code and the performance overhead, we will also prevent ourselves from using some features of each specific database. I believe that's why there's no popular ORM that can do SQL and NoSQL.

Could you give me samples to make me on the same channel?

I think interfaces would not hurt so much, say if we are using postgresql, then the code implementing the interface may be like:

```go
type PostgresDB struct {
  pool pq.Pool
}

func (db *PostgresDB) SaveBlock(block Block) error {
  err := db.pool.Execute('Save Block query', block)
  // other operations
  return err
}

...other implementations
```

And then, the DB implementation for ScyllaDB may be:
```go
type ScyllaDB struct {
  connection scylla.Connection
}

func (db *ScyllaDB) SaveBlock(block Block) error {
  // some operation saving block into ScyllaDB
  return err
}
```

Both of them shared the same interface so we can reuse the code in our implementation codes without any changes:

```go
func (w Woker) Start(ctx Context, queue BlockQueue) {
  for block := range queue.Listen() {
    ...operations
    err := ctx.SaveBlock(block)
    // ..other operations
  }
} 
```

Therefore, from my side, I only see the benefit rather than the disadvantages using interface.

> I think we want to support different inputs and conform them to the same output. I think your approach works for Blocks but not for chain-specific elements. Let's say chains A and B have different proposal data structures, how can we accommodate the Database interface to persist the different shapes of the proposal?

That is the reasons why the design only serves Block to interface and make Juno only responsible for block indexer. The example from you, Chains A and B haves different proposal data, then we should build chain A gov module in A Callisto and in B chain B gov module to serve two cases from my side.

I have an example how Soljuno define a [custom db interface](https://github.com/forbole/soljuno/blob/master/modules/stake/module.go#L23) and [custom client proxy interface](https://github.com/forbole/soljuno/blob/master/modules/stake/module.go#L12) for stake module only for Solana.

So, basically, it would be better to implement module in Callisto to handle the custom things you mentioned.

> That's actually a great point, are we not able to mock/spy/stub the Database client?

Yeah, it is not easy to do that since currently worker does not have interface and everything is only for Cosmos. Say, the returning block is `*types.ResultBlock` inside [node.Block](https://github.com/forbole/juno/blob/cosmos/v0.47.x/node/node.go#L31) method and its interface is too complex as well. And now enqueuer is a function that depends on the [*parsing.Context](https://github.com/forbole/juno/blob/cosmos/v0.47.x/cmd/start/cmd.go#L123) instance, then [start function](https://github.com/forbole/juno/blob/cosmos/v0.47.x/cmd/start/cmd.go#L123) includes many instance structure so we can not test it in a easy way.

Reply:
Regarding the `interface`/generalization of the datastore/underlying database.

Let's say you now want your worker to be able to bulk/batch persist your parsed transaction.

In Posgresql there is no such thing (there's `COPY` but it doesn't fit the example use case). You will end up with either a saving function like `saveBulk(tx []Tx)` or repeating the call to `save(tx Tx)`.

Now with MongoDB use could have used the `bulkWrite` methods so your function signature will be `save(tx []Tx, bulk bool)`, so you could conditionally call `bulkWrite` for your MongoDB implementation of your save method.

But we now have this `bulk bool` param that is only relevant to MongoDB's implementation, therefore we hindered the abstraction of the `save` method.

Another example would be Postgres prepared statements, which can help a bit with performances. Same thing if you want to control the underling implementation from the interface you will have to hinder the abstraction layer you've just created.

---

For the customization of the database repository.

The industry standard for many years now when it comes to the repository pattern is to have one repository per entity. I believe it has passed the test of time.

So we should have `BlockRepository`, `TransactionRepository`, etc with each of them having a `save`, `load`, etc.

I would refrain from having a `Database` with methods `saveBlock`, `saveTransaction`, etc.

---

From the research I have conducted online. Our `Context` struct should only have a reference to the database client.

Reply:
Yeah, you are right. It would be to better to be renamed into `BlockRepository` rather than `Database` name used by legacy Juno to make it clear.